Shared Library Build Broken on Windows

## System info
**Operating System**: Windows 10 x64 1709
**Shaka Packager Version**: 435d69c3a6452cd868a09fa7c48e762004358664

## Issue and steps to reproduce the problem
```
set GYP_DEFINES=target_arch=x64 libpackager_type=shared_library
gclient runhooks
cd src
ninja -C out/Debug_x64
```

What is the expected result?
libpackager should compile as a DLL successfully.

What happens instead?
Errors such as the following occur:
```
[1/47] CXX obj\status.status.obj
FAILED: obj/status.status.obj 
ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\cl.exe" /nologo /showIncludes /FC @obj\status.status.obj.rsp /c ..\..\packager\status.cc /Foobj\status.status.obj /Fdobj\status.cc.pdb 
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2143: syntax error: missing ')' before '('
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2079: 'visibility' uses undefined class 'shaka::__attribute__'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2059: syntax error: ')'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2440: 'initializing': cannot convert from 'const char [8]' to 'int'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): note: There is no context in which this conversion is possible
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2146: syntax error: missing ';' before identifier 'Status'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(111): error C2059: syntax error: 'public'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(113): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(113): error C2365: 'shaka::Status': redefinition; previous definition was 'data variable'
...MORE ERRORS HERE
```

The build breaks. But I got a different erro:

ninja: Entering directory `out/Debug_x64'
[264/661] CXX obj\status.status.obj
FAILED: obj/status.status.obj
ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\cl.exe" /nologo /showIncludes /FC @obj\status.status.obj.rsp /c ..\..\packager\status.cc /Foobj\status.status.obj /Fdobj\status.cc.pdb
c:\users\hmchen\shaka_packager_github\src\packager\status.h(150): error C2220: warning treated as error - no 'object' file generated
c:\users\hmchen\shaka_packager_github\src\packager\status.h(150): warning C4251: 'shaka::Status::error_message_': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'shaka::Status'
c:\program files (x86)\microsoft visual studio 14.0\vc\include\xstring(2633): note: see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>'
[277/661] CXX obj\base\trace_event\base.trace_log.obj
ninja: build stopped: subcommand failed.

--- a/packager/common.gypi
+++ b/packager/common.gypi
@@ -52,6 +52,8 @@
                  # the current code page. It typically happens when compiling
                  # the code in CJK environment if there is non-ASCII characters
                  # in the file.
+          4251,  # Warnings about private std::string in Status in a shared
+                 # library config on Windows.
         ],
       }, {
         # We do not have control over non-shaka code. Disable some warnings to

--- a/packager/file/file.gyp
+++ b/packager/file/file.gyp
@@ -43,6 +43,14 @@
         '../third_party/curl/curl.gyp:libcurl',
         '../version/version.gyp:version',
       ],
+      'conditions': [
+        ['libpackager_type == "shared_library"', {
+          'defines': [
+            'SHARED_LIBRARY_BUILD',
+            'SHAKA_IMPLEMENTATION',
+          ],
+        }],
+      ],
     },
     {
       'target_name': 'file_unittest',

--- a/packager/file/file.h
+++ b/packager/file/file.h
@@ -13,6 +13,7 @@
 
 #include "packager/base/macros.h"
 #include "packager/file/public/buffer_callback_params.h"
+#include "packager/status.h"
 
 namespace shaka {
 
@@ -24,7 +25,7 @@ extern const char* kHttpFilePrefix;
 const int64_t kWholeFile = -1;
 
 /// Define an abstract file interface.
-class File {
+class SHAKA_EXPORT File {
  public:
   /// Open the specified file.
   /// This is a file factory method, it opens a proper file automatically

--- a/packager/packager.gyp
+++ b/packager/packager.gyp
@@ -100,12 +100,16 @@
       ],
       'dependencies': [
         'base/base.gyp:base',
-        'file/file.gyp:file',
         'libpackager',
         'third_party/gflags/gflags.gyp:gflags',
         'tools/license_notice.gyp:license_notice',
       ],
       'conditions': [
+        ['libpackager_type == "shared_library"', {
+          'defines': [
+            'SHARED_LIBRARY_BUILD',
+          ],
+        }],
         ['profiling==1', {
           'dependencies': [
             'base/allocator/allocator.gyp:allocator',
@@ -134,6 +138,13 @@
       'type': '<(gtest_target_type)',
       'sources': [
         'packager_test.cc',
+      ],
+      'conditions': [
+        ['libpackager_type == "shared_library"', {
+          'defines': [
+            'SHARED_LIBRARY_BUILD',
+          ],
+        }],
       ],
       'dependencies': [
         'libpackager',

